edk2toollib/windows/policy/firmware_policy.py

```diff
@@ -0,0 +1,563 @@
+# @file
+#
+# Library to parse/manage/build unsigned firmware policy blobs
+#
+# Copyright (c), Microsoft Corporation
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+
+##
+# A firmware policy is conceptually a set of key value pair "rules".
+# Rules can be 1-time actions, states persistent while the policy is in effect,
+# or targeting information describing the machine where the policy is applicable
+#
+# Each key is composed of a UINT32 RootKey followed by String SubKeyName & ValueName
+# For example:
+#   RootKey = 0xEF100000
+#   SubKeyName = "PolicyGroupFoo"
+#   ValueName = "TpmClear"
+#   MyExampleKey = EF100000\PolicyGroupFoo\TpmClear
+#
+# Each value has a type that can be primitive or a variable-length structure.
+#
+##
+# The policy binary, pack(1), little endian
+##
+# UINT16   FormatVersion
+# UINT32   PolicyVersion
+# GUID     PolicyPublisher
+# UINT16   Reserved1Count  # 0
+#          Reserved1[Reserved1Count]   # not present
+# UINT32   OptionFlags     # 0
+# UINT16   Reserved2Count  # 0
+# UINT16   RulesCount
+#          Reserved2[Reserved2Count]   # not present
+# RULE     Rules[RulesCount]  # inline, not a pointer, see class Rule below
+# BYTE     ValueTable[]  #inline, not a pointer
+#
+##
+# The RULE structure is as follows
+##
+# UINT32 RootKey
+# UINT32 OffsetToSubKeyName  # ValueTable offset
+# UINT32 OffsetToValueName   # ValueTable offset
+# UINT32 OffsetToValue       # ValueTable offset to a PolicyValueType + PolicyValue
+#
+##
+# Each Rule corresponds to 3 entries in the ValueTable
```

Comment:
should add tests for the things below.  Especially interesting is the reuse of subkeys

Reply:
firmware_policy.py serialize functions do **not** currently perform this optimization, but some other libraries generate them.  The FromFileStream() methods _have_ been tested to work against several optimized policies.

- In your opinion, should we add this optimization as an option to the FirmwarePolicy Serialize() method?
- if yes, then we can build the test cases using this library.  If no, then I'll locate some binary blobs created by other libraries.

Reply:
no need to add optimization....but maybe a unit test with existing policy just to make sure that functionality doesn't break.

Reply:
The upcoming commit will add a test case and a large, optimized policy to be tested.

Reply:
Committed in e89edefe55dac2101829619ffcc1c2b1626ab2e8